src/shred_network/collector/duplicate_shred_handler.zig

```diff
@@ -0,0 +1,354 @@
+            });
+        }
+
+        return try chunks.toOwnedSlice(allocator);
+    }
+};
+
+fn hasDuplicateForSlot(
+    gossip_table_rw: *sig.sync.RwMux(sig.gossip.GossipTable),
+    my_pubkey: sig.core.Pubkey,
+    slot: Slot,
+) bool {
+    var gossip_table, var lock = gossip_table_rw.readWithLock();
+    defer lock.unlock();
+
+    if (gossip_table.pubkey_to_values.get(my_pubkey)) |records| {
+        for (records.keys()) |record_ix| {
+            const versioned_data = gossip_table.store.getByIndex(record_ix);
+            switch (versioned_data.data) {
+                .DuplicateShred => |dup| {
+                    _, const dup_shred = dup;
+                    if (dup_shred.slot == slot) return true;
+                },
+                else => {},
+            }
+        }
+    }
+    return false;
+}
+
+pub fn computeRingOffset(
+    gossip_table_rw: *sig.sync.RwMux(sig.gossip.GossipTable),
+    my_pubkey: sig.core.Pubkey,
+) u16 {
+    const MAX_DUPLICATE_SHREDS = sig.gossip.data.MAX_DUPLICATE_SHREDS;
+    var num_dup_shreds: u16 = 0;
+    var oldest_index: u16 = 0;
+    var maybe_oldest_wallclock: ?u64 = null;
+
+    var gossip_table, var lock = gossip_table_rw.readWithLock();
+    defer lock.unlock();
+
+    if (gossip_table.pubkey_to_values.get(my_pubkey)) |records| {
+        for (records.keys()) |record_ix| {
+            const versioned_data = gossip_table.store.getByIndex(record_ix);
+            switch (versioned_data.data) {
+                .DuplicateShred => |dup| {
+                    const index, const dup_shred = dup;
+                    num_dup_shreds +%= 1;
+                    const wc = dup_shred.wallclock;
+                    if (maybe_oldest_wallclock) |old_wc| {
+                        if (wc < old_wc or (wc == old_wc and index < oldest_index)) {
+                            maybe_oldest_wallclock = wc;
+                            oldest_index = index;
+                        }
+                    } else {
+                        maybe_oldest_wallclock = wc;
+                        oldest_index = index;
+                    }
+                },
+                else => {},
+            }
+        }
+    }
+
+    return if (num_dup_shreds < MAX_DUPLICATE_SHREDS) num_dup_shreds else oldest_index;
+}
+
+pub fn serializeDuplicateProof(
+    allocator: std.mem.Allocator,
+    shred_payload: []const u8,
+    other_payload: []const u8,
+) ![]const u8 {
+    // TODO validate the shreds. Implement check_shreds in Agave
+    const proof = sig.ledger.meta.DuplicateSlotProof{
+        .shred1 = shred_payload,
+        .shred2 = other_payload,
+    };
+    var proof_data: std.ArrayListUnmanaged(u8) = .empty;
+    errdefer proof_data.deinit(allocator);
+    try bincode.write(proof_data.writer(allocator), proof, bincode.Params.standard);
+    const bytes = try proof_data.toOwnedSlice(allocator);
+    return bytes;
+}
+
+fn enqueueDuplicateShredCrdsValues(
+    gossip: *sig.gossip.GossipService,
```

Comment:
depending on the entire gossip service struct? do we do that anywhere else in the codebase?